File: Python/313__Super_Ugly_Number.py
class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        """
        :type n: int
        :type primes: List[int]
        :rtype: int
        """
        ugly_numbers = [0] * n
        ugly_numbers[0] = 1
        indexes = [0] * len(primes)
        refactors = [i for i in primes]
        for i in range(1, n):
            ugly_numbers[i] = min(refactors)
            for ind in [k for k, x in enumerate(refactors) if x == ugly_numbers[i]]:
                indexes[ind] += 1
                refactors[ind] = ugly_numbers[indexes[ind]]*primes[ind]
        return ugly_numbers[-1]

File: Java/313__Super_Ugly_Number.java
class Solution {
    public int nthSuperUglyNumber(int n, int[] primes) {
        int[][] save = new int[primes.length][n + 1];
        int[] result = new int[n];
        for (int i = 0; i < primes.length; i++)
            save[i][0] = 1;
        int[] pointer = new int[primes.length];
        for (int i = 1; i <= n; i++) {
            int min = Integer.MAX_VALUE;
            for (int j = 0; j < primes.length; j++)
                min = Math.min(save[j][pointer[j]], min);
            for (int j = 0; j < primes.length; j++) {
                if (save[j][pointer[j]] == min)
                    pointer[j]++;
                save[j][i] = min * primes[j];
            }
            result[i - 1] = min;
        }
        return result[n - 1];
    }
}
